Adds "welder crafting" to iron sheets, rods and tiles.

GuillaumePrata opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
This adds a new way to craft these 3 items without having to open the crafting menu by using a welder on them with left or right clicks.

Rods: Left click is the old behavior of crafting them back into sheets, right click turns them into floor tiles.
Floor tiles: Left click goes back to sheets, right click craft rods.
Iron sheets: Left click craft rods, right click craft floor tiles

They have contextual screen tips and balloon alerts.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
I keep seeing people complain about engineers not being able to fix breaches efficiently without RCDs and to a certain point they are right.
Fixing rooms without a RCD sucks, even for me and I hate RCDs.

We have 2 options:
Foam grenades to quickly patch the room, in theory, in practice it is a slow process because engineers only have access to the "smart" version of foam grenades that leave the room open. A noop trap considering that metal foam can't fix breaches under solid objects like consoles and lockers.
Manual construction with rods and metal tiles, it is not only slower but it hogs all your backpack as you need to carry rods, tiles and metal, then keep swapping between each of them and using the crafting menu if more are needed.

There is a third option that is the RAT and I did stop working on this when the Forklift PR was open, now that it was closed I decided to come back and possibly try other changes to improve our base construction system while we wait.

Now with this idea of "welder crafting" you can keep metal sheets (or rods/floor tiles) on your backpack and just click on it with a welder to smoothly make rods and floor tiles with left/right click as you need them without dealing with a huge TGUI crafting menu blocking a third of your screen.
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: Guillaume Prata
qol: You can now "welder craft" iron sheets/rods/floor tiles by Left or Right clicking one of these items with a welder. It will craft one of the other two options depending on the Left/Right click. There is a contextual screen tip too so just hover over them with a welder for details on the results.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

GuillaumePrata committed:
Adds "welder crafting" to iron sheets, rods and tiles.

GuillaumePrata committed:
Balloon alerts and swaps broken `put_in_hands` with functional `put_in_inactive_hand`

JohnFulpWillard commented on code/game/objects/items/stacks/rods.dm:
```diff
@@ -60,30 +67,53 @@ GLOBAL_LIST_INIT(rod_recipes, list ( \
 
-/obj/item/stack/rods/attackby(obj/item/W, mob/user, params)
-	if(W.tool_behaviour == TOOL_WELDER)
+/obj/item/stack/rods/attackby(obj/item/attackby_item, mob/user, params)
```
Please use ``welder_act`` instead

JohnFulpWillard commented on code/game/objects/items/stacks/rods.dm:
```diff
@@ -60,30 +67,53 @@ GLOBAL_LIST_INIT(rod_recipes, list ( \
 			return
-
-		if(W.use_tool(src, user, 0, volume=40))
+		if(attackby_item.use_tool(src, user, 0, volume = 40))
```
```suggestion
		if(attackby_item.use_tool(src, user, delay = 0, volume = 40))
```

JohnFulpWillard commented on code/game/objects/items/stacks/rods.dm:
```diff
@@ -60,30 +67,53 @@ GLOBAL_LIST_INIT(rod_recipes, list ( \
+				ignored_mobs = user
+			)
+			var/obj/item/stack/rods/welded_rod = src
 			src = null
```
wtf is this can you please get rid of it

JohnFulpWillard commented on code/game/objects/items/stacks/rods.dm:
```diff
@@ -60,30 +67,53 @@ GLOBAL_LIST_INIT(rod_recipes, list ( \
 	else
 		return ..()
 
+/obj/item/stack/rods/attackby_secondary(obj/item/attackby_item, mob/user, params)
```
``welder_act_secondary()`` please

GuillaumePrata replied:
I had no idea that even existed but it made things a lot easier for me, thonks!

JohnFulpWillard commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -173,6 +183,43 @@ GLOBAL_LIST_INIT(metal_recipes, list ( \
 	user.visible_message(span_suicide("[user] begins whacking [user.p_them()]self over the head with \the [src]! It looks like [user.p_theyre()] trying to commit suicide!"))
 	return BRUTELOSS
 
+/obj/item/stack/sheet/iron/attackby(obj/item/attackby_item, mob/user, params)
```
same here, ``welder_act()`` and ``welder_act_secondary()``

JohnFulpWillard commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -173,6 +183,43 @@ GLOBAL_LIST_INIT(metal_recipes, list ( \
+/obj/item/stack/sheet/iron/attackby(obj/item/attackby_item, mob/user, params)
+	if(attackby_item.tool_behaviour == TOOL_WELDER)
+		if(attackby_item.use_tool(src, user, 0, volume = 40))
+			var/obj/item/stack/rods/two/new_item = new(usr.loc)
```
```suggestion
			var/obj/item/stack/rods/two/new_item = new(user.loc)
```

JohnFulpWillard commented on code/game/objects/items/stacks/tiles/tile_iron.dm:
```diff
@@ -83,23 +83,65 @@
+	)
+	AddElement(/datum/element/contextual_screentip_tools, tool_behaviors)
+
+/obj/item/stack/tile/iron/attackby(obj/item/attackby_item, mob/user, params)
```
i think you get it by now

GuillaumePrata committed:
Changes `attack_by` to `welder_act` and some small stuff

Like better indentation, killing `src = null` and else/returns

GuillaumePrata committed:
Moves the return to the proper place

Ye I lied about better indentation on my last commit

JohnFulpWillard commented on code/game/objects/items/stacks/rods.dm:
```diff
@@ -60,30 +67,46 @@ GLOBAL_LIST_INIT(rod_recipes, list ( \
+		return
+	if(tool.use_tool(src, user, delay = 0, volume = 40))
+		var/obj/item/stack/sheet/iron/new_item = new(user.loc)
+		balloon_alert(user, "crafting sheets...")
```
there's no need for a balloon alert, we can already tell the activity was a success from the visible message and the rods now being in your hands, since this has no delay.

JohnFulpWillard commented on code/game/objects/items/stacks/rods.dm:
```diff
@@ -60,30 +67,46 @@ GLOBAL_LIST_INIT(rod_recipes, list ( \
+			vision_distance = COMBAT_MESSAGE_RANGE,
+			ignored_mobs = user
+		)
+		var/obj/item/stack/rods/welded_rod = src
+		welded_rod.use(2)
```
```suggestion
		use(2)
```

JohnFulpWillard reviewed (commented):
ditto for all

GuillaumePrata committed:
Die `balloon_alert` and better use of `use()` too

Willard used the same reasoning I used to kill ballon alerts in other PR I made, how the fuck didn't I think of this myself????
